src/JustSaying/Messaging/Channels/ConsumerBus.cs

```diff
@@ -0,0 +1,75 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using JustSaying.AwsTools.MessageHandling;
+using Microsoft.Extensions.Logging;
+
+namespace JustSaying.Messaging.Channels
+{
+    internal interface IConsumerBus
+    {
+        void Start(CancellationToken stoppingToken);
+        Task Completion { get; }
```

Comment:
There's two tasks here - one to ensure the bus is started and is in a ready state to continue, and the other is to wait for it to complete. Callers should be able to wait for completion on shutdown so they can complete gracefully, and they should also be able to wait until the bus is started before continuing program startup - both optional of course. We're following the same pattern that `Channels` uses to notify completion, while also allowing async startup of the queue, as we have to wait for an async lock.

Reply:
_Ponders further_ - just seems weird to me to have an interface with a property of `Task` on it.

What if you waited on the task before it started? You'd have grabbed a reference to a different task? Also it could be null, which would be weird and cause a NullReferenceException if you tried to await it without checking.

Reply:
This pattern felt a little weird to me at first too, but I've come around to it a bit after using it in channels: https://docs.microsoft.com/en-us/dotnet/api/system.threading.channels.channelreader-1.completion?view=netcore-3.1#System_Threading_Channels_ChannelReader_1_Completion

We can protect this a bit by throwing invalid op if this task is awaited before calling Start. Also, this is never called from user code, it's internal and only used by `IJustSayingBus` so we just need to make sure our usage is correct 🙂 .

Reply:
Looking at this again, we perhaps don't need to expose this pattern to users of `IConsumerBus` - we don't really have a need to wait until the bus is started here, so we could just return the completion task from `Start`.

Reply:
This has been resolved, and the IConsumerBus.Start now returns the completion task.